[jit][tensorexpr] Added aten::batch_norm into fuser when in inference mode

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54204 [jit][tensorexpr] Added aten::batch_norm into fuser when in inference mode**

Differential Revision: [D27134348](https://our.internmc.facebook.com/intern/diff/D27134348)

huiguoo committed:
[jit][tensorexpr] Added aten::batch_norm into fuser when in inference mode

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1062,6 +1063,77 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
+
+            // parameter list: input, weight, bias, mean, var, training,
+            // momentum, eps, cudnn_enabled
+            Tensor* input = tensors_.at(n->inputs()[0]->unique());
```
Nit: use `n->input(X)` instead of `n->inputs()[X]`.

ZolotukhinM commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1062,6 +1063,77 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
+              auto output = input->call(axes) * alpha - mean->call(c) * alpha;
+              return demoteOutput(output, n->output());
+            } else {
+              Tensor* weight = tensors_.at(n->inputs()[1]->unique());
+              Tensor* bias = tensors_.at(n->inputs()[2]->unique());
+
+              auto inv_var = rsqrt(var->call(c) + eps);
+              auto weight_v = weight->call(c);
+              auto bias_v = bias->call(c);
+              auto alpha = inv_var * weight_v;
+              auto beta = bias_v - mean->call(c) * alpha;
+              auto output = input->call(axes) * alpha + beta;
+              return demoteOutput(output, n->output());
```
I wonder if we could merge these four cases using `weight_v = 1` and `bias_v = 0` as default values.

ZolotukhinM reviewed (approved):
Looks good, a couple of comments inline!

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1062,6 +1063,77 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
+            ExprHandle eps = constant(n->inputs()[7]);
+
+            // axes: N, C, H, W
+            std::vector<VarHandle> c(axes.begin() + 1, axes.begin() + 2);
```
You're just getting one axis here, right (c)?  I think I'd be inclined to do `VarHandle c = axes[1];` here instead of constructing a one-element vector.  `call` has a variadic overload so nothing below should have to change (at least, I don't think so)

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1062,6 +1063,77 @@ Tensor* TensorExprKernel::computeValue(const torch::jit::Value* v) {
+              auto weight_v = weight->call(c);
+              auto alpha = inv_var * weight_v;
+              auto output = input->call(axes) * alpha - mean->call(c) * alpha;
+              return demoteOutput(output, n->output());
```
I wonder if we need some promoteInputs as well.  I'm not certain here b/c pytorch type promotion is complicated, but I wonder what you get if, e.g., `input` is a `FloatTensor` and `mean` is a `DoubleTensor`.  (Not something that's likely to happen but we should handle it correctly all the same).

bertmaher reviewed (commented):
Looks nice!  Can you check two things before landing:
1. How does that `training` boolean parameter get set?  Is it from the `eval()` state of the model?  What happens if the user switches the model between `eval()` and `train()`?
2. The promoteInputs question I have below; make sure we "do the right thing" if the argument types don't match.

huiguoo commented:
@bertmaher 
> How does that training boolean parameter get set? Is it from the eval() state of the model? What happens if the user switches the model between eval() and train()? 

Great question! Has investigated on this and below is what I found. 
There are two types of calls to `aten::batch_norm` in a pytorch program : 
(1) `nn.functional.batch_norm(input, mean, var, training=false)`
(2) `m = nn.BatchNorm2d(c)
       output=m(input)`
 In the first case, `training` parameter in `aten::batch_norm` is a constant and set to false in default. The state change between eval() and train() doesn't affect its value. We'll check the value of `training` parameter and only fuse batch_norm when `training` is false.
 In the second case, the `training` parameter is determined by the model state: `false` in eval() and `true` in train(). If the user switches the model between eval() and train(), the value of `training` parameter will change correspondingly. In the torchscript graph, `training` is a `prim::GetAttr` node - we would not be able to fuse batch_norm unless we freeze the graph first so `training` becomes a constant.

huiguoo committed:
Update on "[jit][tensorexpr] Added aten::batch_norm into fuser when in inference mode"


Differential Revision: [D27134348](https://our.internmc.facebook.com/intern/diff/D27134348)

[ghstack-poisoned]

huiguoo committed:
Update on "[jit][tensorexpr] Added aten::batch_norm into fuser when in inference mode"


Differential Revision: [D27134348](https://our.internmc.facebook.com/intern/diff/D27134348)

[ghstack-poisoned]

huiguoo committed:
Update on "[jit][tensorexpr] Added aten::batch_norm into fuser when in inference mode"


Differential Revision: [D27134348](https://our.internmc.facebook.com/intern/diff/D27134348)

[ghstack-poisoned]

huiguoo committed:
Update on "[jit][tensorexpr] Added aten::batch_norm into fuser when in inference mode"


Differential Revision: [D27134348](https://our.internmc.facebook.com/intern/diff/D27134348)

[ghstack-poisoned]

huiguoo committed:
Update on "[jit][tensorexpr] Added aten::batch_norm into fuser when in inference mode"


Differential Revision: [D27134348](https://our.internmc.facebook.com/intern/diff/D27134348)

[ghstack-poisoned]
